---
title: "README"
author: "Charlotte Barclay"
date: "26/01/2022"
output: html_document
---

# *Homiphora californensis* - a positive control

The following document describes and provides the code for a reproduction of the assembly for H. californensis as in the study.The H. californensis paper [here](https://academic.oup.com/g3journal/article/11/11/jkab302/6358137) used the wtdbg2 assembler, with Arrow and Pilon polishing to generate their assembly. This was followed with HiRise scaffolding with the HiC samples. 

In this reproduction, we will reproduce the same draft assembly steps, followed by SALSA - to see if the same/comparable results are achieved.

## Running assembly and generating a consensus sequence

#### 1. Prepare conda environment

The draft assembly was generated with wtdbg2 assembler, Arrow and Pilon. These have been installed into an existing conda environment which already has SALSA and relevant pre-requisites installed.

Install into 'condahic' environment using the following code: 

```{bash, eval=FALSE}
conda activate condahic
conda install -c bioconda wtdbg
conda install -c bioconda arrow
conda install -c bioconda pilon
```

#### 2. Run assembly and generate consensus

Estimate 15-30mins. Wtdbg2 has two key components: an assembler **wtdbg2** and a consenser **wtpoa-cns**. 

Executable wtdbg2 assembles raw reads and generates the contig layout and edge sequences in a file "prefix.ctg.lay.gz". Executable wtpoa-cns takes this file as input and produces the final consensus in FASTA. A typical workflow looks like this:

```{bash, eval=FALSE}
./wtdbg2 -x rs -g 4.6m -t 16 -i reads.fa.gz -fo prefix
./wtpoa-cns -t 16 -i prefix.ctg.lay.gz -fo prefix.ctg.fa
```

***Parameters*** 

1. -g is the estimated genome size
2. -x specifies the sequencing technology (value "rs" for PacBio RSII, "sq" for PacBio Sequel, "ccs" for PacBio CCS reads and "ont" for Oxford Nanopore)
3. -i indicates the reads (need to repeat -i before each file as otherwise it only uses first file provided).


```{basH, eval=FALSE}
#ASSEMBLY

##wtdbg2 -x sq -g 156m -t 16 -i reads.fa.gz -fo prefix

wtdbg2 -x sq -g 98m -t 16 -i SRR10237148.fasta -i SRR10237149.fasta -i SRR10237137.fasta -fo Hcaliforensis

#CONSENSUS

##wtpoa-cns -t 16 -i prefix.ctg.lay.gz -fo prefix.ctg.fa

wtpoa-cns -t 16 -i Hcaliforensis.ctg.lay.gz -fo Hcaliforensis.ctg.fa


pscp -P 22 C:\Users\cbarc\Documents\LAB_OriginsOfMulticellularity\Data\HormiphoraCaliforensis\*.fasta.gz cbarcl01@drake.ubc.ca:/home/cbarclay/Documents/Hcaliforensis/pacBio

pscp -P 22 C:\Users\cbarc\Documents\LAB_OriginsOfMulticellularity\Data\HormiphoraCaliforensis\HiC\*.fasta.gz cbarcl01@drake.ubc.ca:/home/cbarclay/Documents/Hcaliforensis/data

pscp -P 22 cbarclay@drake.phas.ubc.ca:/home/cbarclay/Documents/Mleidyi/workflow/scaffolds/salsa_scaffolds.hic C:\Users\cbarc\Documents\LAB_OriginsOfMulticellularity\Data\MyData

```

#### 3. Create an index

```{bash eval=FALSE}
bowtie2-build /home/cbarclay/Documents/Hcaliforensis/pacBio/Hcaliforensis.ctg.fasta /home/cbarclay/Documents/Hcaliforensis/index
```

```{bash, eval=FALSE}
bowtie2 -f -x /home/cbarclay/Documents/Hcaliforensis/index/index.fai \ -1 /home/cbarclay/Documents/Hcaliforensis/data/SRR10237146.fasta \ -2 /home/cbarclay/Documents/Hcaliforensis/data/SRR10237147.fasta \ -S /home/cbarclay/Documents/Hcaliforensis/data/hcal_hic.sam
```
